syntax = "proto2";

message SockAddrIn {
  required uint64 addr = 1;
  required uint32 port = 2;
}

message MergeRequest {
  required bytes object = 1;
  optional uint64 epoch = 2;
}

message TxnRequest {
  required uint64 request_id = 1;
  required bytes txn = 2;
}

message ForwardedTxnRequest {
  required uint64 request_id = 1;
  required bytes txn = 2;
  required SockAddrIn src_addr = 3;
}

message TxnReply {
  required uint64 request_id = 1;
  required bytes reply = 2;
}

message Prepare {
  required uint64 txn_index = 1;
  required bytes txn = 2;
  required uint64 num_committed = 3;
}

message PrepareOk {
  required uint64 txn_index = 1;
  required uint64 replica_index = 2;
}

message SyncRequest {
  required uint64 replica_index = 1;
  required uint64 epoch = 2;
  optional bytes old_object = 3;
}

message SyncReply {
  required uint64 replica_index = 1;
  required uint64 epoch = 2;
  required bytes object = 3;
}

message Start {
  required bytes object = 1;
  required uint64 epoch = 2;
}

message StartAck {
  required uint64 replica_index = 1;
  required uint64 epoch = 2;
}

message ServerMessage {
  oneof msg {
    MergeRequest merge_request = 1;
    TxnRequest txn_request = 2;
    ForwardedTxnRequest forwarded_txn_request = 3;
    TxnReply txn_reply = 4;
    Prepare prepare = 5;
    PrepareOk prepare_ok = 6;
    SyncRequest sync_request = 7;
    SyncReply sync_reply = 8;
    Start start = 9;
    StartAck start_ack = 10;
  }
}
